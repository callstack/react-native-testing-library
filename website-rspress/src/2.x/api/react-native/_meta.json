[
  {
    "type": "file",
    "name": "index",
    "label": "Table of Contents"
  },
  {
    "type": "dir",
    "name": "classes",
    "label": "Classes"
  },
  {
    "type": "dir",
    "name": "interfaces",
    "label": "Interfaces"
  },
  {
    "type": "dir",
    "name": "types",
    "label": "Types"
  }
]
