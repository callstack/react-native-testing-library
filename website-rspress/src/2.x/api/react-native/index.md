# Re.Pack React Native APIs

## Table of contents

### Classes

- [Chunk](./classes/Chunk.md)
- [ChunkManager](./classes/ChunkManager.md)

### Interfaces

- [ChunkConfig](./interfaces/ChunkConfig.md)
- [ChunkManagerConfig](./interfaces/ChunkManagerConfig.md)
- [RemoteChunkLocation](./interfaces/RemoteChunkLocation.md)
- [StorageApi](./interfaces/StorageApi.md)

### Type aliases

- [RemoteChunkResolver](./types/RemoteChunkResolver.md)
