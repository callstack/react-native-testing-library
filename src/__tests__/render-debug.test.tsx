/* eslint-disable no-console */
import * as React from 'react';
import { View, Text, TextInput, Pressable } from 'react-native';
import stripAnsi from 'strip-ansi';
import { render, fireEvent, configure } from '..';

type ConsoleLogMock = jest.Mock<Array<string>>;

const PLACEHOLDER_FRESHNESS = 'Add custom freshness';
const PLACEHOLDER_CHEF = 'Who inspected freshness?';
const INPUT_FRESHNESS = 'Custom Freshie';
const INPUT_CHEF = 'I inspected freshie';
const DEFAULT_INPUT_CHEF = 'What did you inspect?';
const DEFAULT_INPUT_CUSTOMER = 'What banana?';

const ignoreWarnings = ['Using debug("message") is deprecated'];

const realConsoleWarn = console.warn;

beforeEach(() => {
  jest.spyOn(console, 'log').mockImplementation(() => {});
  jest.spyOn(console, 'warn').mockImplementation((message) => {
    if (!ignoreWarnings.some((warning) => message.includes(warning))) {
      realConsoleWarn(message);
    }
  });
});

afterEach(() => {
  jest.restoreAllMocks();
});

class MyButton extends React.Component<any> {
  render() {
    return (
      <Pressable onPress={this.props.onPress}>
        <Text>{this.props.children}</Text>
      </Pressable>
    );
  }
}

class Banana extends React.Component<any, { fresh: boolean }> {
  state = {
    fresh: false,
  };

  componentDidUpdate() {
    if (this.props.onUpdate) {
      this.props.onUpdate();
    }
  }

  componentWillUnmount() {
    if (this.props.onUnmount) {
      this.props.onUnmount();
    }
  }

  changeFresh = () => {
    this.setState((state) => ({
      fresh: !state.fresh,
    }));
  };

  render() {
    const test = 0;
    return (
      <View>
        <Text>Is the banana fresh?</Text>
        <Text testID="bananaFresh">{this.state.fresh ? 'fresh' : 'not fresh'}</Text>
        <TextInput
          testID="bananaCustomFreshness"
          placeholder={PLACEHOLDER_FRESHNESS}
          value={INPUT_FRESHNESS}
        />
        <TextInput
          testID="bananaChef"
          placeholder={PLACEHOLDER_CHEF}
          value={INPUT_CHEF}
          defaultValue={DEFAULT_INPUT_CHEF}
        />
        <TextInput defaultValue={DEFAULT_INPUT_CUSTOMER} />
        <TextInput defaultValue={'hello'} value="" />
        <MyButton onPress={this.changeFresh} type="primary">
          Change freshness!
        </MyButton>
        <Text testID="duplicateText">First Text</Text>
        <Text testID="duplicateText">Second Text</Text>
        <Text>{test}</Text>
      </View>
    );
  }
}

test('debug', () => {
  const { debug } = render(<Banana />);

  debug();
  debug('my custom message');
  debug.shallow();
  debug.shallow('my other custom message');
  debug({ message: 'another custom message' });

  const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
  expect(stripAnsi(mockCalls[1][0] + mockCalls[1][1])).toMatchSnapshot('with message');
  expect(stripAnsi(mockCalls[2][0])).toMatchSnapshot('shallow');
  expect(stripAnsi(mockCalls[3][0] + mockCalls[3][1])).toMatchSnapshot('shallow with message');
  expect(stripAnsi(mockCalls[4][0] + mockCalls[4][1])).toMatchSnapshot('another custom message');

  const mockWarnCalls = (console.warn as any as ConsoleLogMock).mock.calls;
  expect(mockWarnCalls[0]).toEqual([
    'Using debug("message") is deprecated and will be removed in future release, please use debug({ message; "message" }) instead.',
  ]);
});

test('debug changing component', () => {
  const { UNSAFE_getByProps, debug } = render(<Banana />);
  fireEvent.press(UNSAFE_getByProps({ type: 'primary' }));

  debug();

  const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot(
    'bananaFresh button message should now be "fresh"'
  );
});

test('debug with only children prop', () => {
  const { debug } = render(<Banana />);
  debug({ mapProps: () => ({}) });

  const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
});

test('debug with only prop whose value is bananaChef', () => {
  const { debug } = render(<Banana />);
  debug({
    mapProps: (props) => {
      const filterProps: Record<string, unknown> = {};
      Object.keys(props).forEach((key) => {
        if (props[key] === 'bananaChef') {
          filterProps[key] = props[key];
        }
      });
      return filterProps;
    },
  });

  const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
});

test('debug with only props from TextInput components', () => {
  const { debug } = render(<Banana />);
  debug({
    mapProps: (props, node) => (node.type === 'TextInput' ? props : {}),
  });

  const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
});

test('debug should use debugOptions from config when no option is specified', () => {
  configure({ defaultDebugOptions: { mapProps: () => ({}) } });

  const { debug } = render(
    <View style={{ backgroundColor: 'red' }}>
      <Text>hello</Text>
    </View>
  );
  debug();

  const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
});

test('filtering out props through mapProps option should not modify component', () => {
  const { debug, getByTestId } = render(<View testID="viewTestID" />);
  debug({ mapProps: () => ({}) });

  expect(getByTestId('viewTestID')).toBeTruthy();
});

test('debug should use given options over config debugOptions', () => {
  configure({ defaultDebugOptions: { mapProps: () => ({}) } });

  const { debug } = render(
    <View style={{ backgroundColor: 'red' }}>
      <Text>hello</Text>
    </View>
  );
  debug({ mapProps: (props) => props });

  const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
});
