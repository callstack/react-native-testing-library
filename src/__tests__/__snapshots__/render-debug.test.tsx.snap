// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug 1`] = `
"<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text
    testID="bananaFresh"
  >
    not fresh
  </Text>
  <TextInput
    placeholder="Add custom freshness"
    testID="bananaCustomFreshness"
    value="Custom Freshie"
  />
  <TextInput
    defaultValue="What did you inspect?"
    placeholder="Who inspected freshness?"
    testID="bananaChef"
    value="I inspected freshie"
  />
  <TextInput
    defaultValue="What banana?"
  />
  <TextInput
    defaultValue="hello"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function onBlur]}
    onClick={[Function onClick]}
    onFocus={[Function onFocus]}
    onResponderGrant={[Function onResponderGrant]}
    onResponderMove={[Function onResponderMove]}
    onResponderRelease={[Function onResponderRelease]}
    onResponderTerminate={[Function onResponderTerminate]}
    onResponderTerminationRequest={[Function onResponderTerminationRequest]}
    onStartShouldSetResponder={[Function onStartShouldSetResponder]}
  >
    <Text>
      Change freshness!
    </Text>
  </View>
  <Text
    testID="duplicateText"
  >
    First Text
  </Text>
  <Text
    testID="duplicateText"
  >
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug changing component: bananaFresh button message should now be "fresh" 1`] = `
"<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text
    testID="bananaFresh"
  >
    fresh
  </Text>
  <TextInput
    placeholder="Add custom freshness"
    testID="bananaCustomFreshness"
    value="Custom Freshie"
  />
  <TextInput
    defaultValue="What did you inspect?"
    placeholder="Who inspected freshness?"
    testID="bananaChef"
    value="I inspected freshie"
  />
  <TextInput
    defaultValue="What banana?"
  />
  <TextInput
    defaultValue="hello"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function onBlur]}
    onClick={[Function onClick]}
    onFocus={[Function onFocus]}
    onResponderGrant={[Function onResponderGrant]}
    onResponderMove={[Function onResponderMove]}
    onResponderRelease={[Function onResponderRelease]}
    onResponderTerminate={[Function onResponderTerminate]}
    onResponderTerminationRequest={[Function onResponderTerminationRequest]}
    onStartShouldSetResponder={[Function onStartShouldSetResponder]}
  >
    <Text>
      Change freshness!
    </Text>
  </View>
  <Text
    testID="duplicateText"
  >
    First Text
  </Text>
  <Text
    testID="duplicateText"
  >
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug should use debugOptions from config when no option is specified 1`] = `
"<View>
  <Text>
    hello
  </Text>
</View>"
`;

exports[`debug should use given options over config debugOptions 1`] = `
"<View
  style={
    {
      "backgroundColor": "red",
    }
  }
>
  <Text>
    hello
  </Text>
</View>"
`;

exports[`debug with only children prop 1`] = `
"<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text>
    not fresh
  </Text>
  <TextInput />
  <TextInput />
  <TextInput />
  <TextInput />
  <View>
    <Text>
      Change freshness!
    </Text>
  </View>
  <Text>
    First Text
  </Text>
  <Text>
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug with only prop whose value is bananaChef 1`] = `
"<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text>
    not fresh
  </Text>
  <TextInput />
  <TextInput
    testID="bananaChef"
  />
  <TextInput />
  <TextInput />
  <View>
    <Text>
      Change freshness!
    </Text>
  </View>
  <Text>
    First Text
  </Text>
  <Text>
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug with only props from TextInput components 1`] = `
"<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text>
    not fresh
  </Text>
  <TextInput
    placeholder="Add custom freshness"
    testID="bananaCustomFreshness"
    value="Custom Freshie"
  />
  <TextInput
    defaultValue="What did you inspect?"
    placeholder="Who inspected freshness?"
    testID="bananaChef"
    value="I inspected freshie"
  />
  <TextInput
    defaultValue="What banana?"
  />
  <TextInput
    defaultValue="hello"
    value=""
  />
  <View>
    <Text>
      Change freshness!
    </Text>
  </View>
  <Text>
    First Text
  </Text>
  <Text>
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug: another custom message 1`] = `
"another custom message

<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text
    testID="bananaFresh"
  >
    not fresh
  </Text>
  <TextInput
    placeholder="Add custom freshness"
    testID="bananaCustomFreshness"
    value="Custom Freshie"
  />
  <TextInput
    defaultValue="What did you inspect?"
    placeholder="Who inspected freshness?"
    testID="bananaChef"
    value="I inspected freshie"
  />
  <TextInput
    defaultValue="What banana?"
  />
  <TextInput
    defaultValue="hello"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function onBlur]}
    onClick={[Function onClick]}
    onFocus={[Function onFocus]}
    onResponderGrant={[Function onResponderGrant]}
    onResponderMove={[Function onResponderMove]}
    onResponderRelease={[Function onResponderRelease]}
    onResponderTerminate={[Function onResponderTerminate]}
    onResponderTerminationRequest={[Function onResponderTerminationRequest]}
    onStartShouldSetResponder={[Function onStartShouldSetResponder]}
  >
    <Text>
      Change freshness!
    </Text>
  </View>
  <Text
    testID="duplicateText"
  >
    First Text
  </Text>
  <Text
    testID="duplicateText"
  >
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug: shallow 1`] = `
"<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text
    testID="bananaFresh"
  >
    not fresh
  </Text>
  <TextInput
    placeholder="Add custom freshness"
    testID="bananaCustomFreshness"
    value="Custom Freshie"
  />
  <TextInput
    defaultValue="What did you inspect?"
    placeholder="Who inspected freshness?"
    testID="bananaChef"
    value="I inspected freshie"
  />
  <TextInput
    defaultValue="What banana?"
  />
  <TextInput
    defaultValue="hello"
    value=""
  />
  <MyButton
    onPress={[Function changeFresh]}
    type="primary"
  >
    Change freshness!
  </MyButton>
  <Text
    testID="duplicateText"
  >
    First Text
  </Text>
  <Text
    testID="duplicateText"
  >
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug: shallow with message 1`] = `
"my other custom message

<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text
    testID="bananaFresh"
  >
    not fresh
  </Text>
  <TextInput
    placeholder="Add custom freshness"
    testID="bananaCustomFreshness"
    value="Custom Freshie"
  />
  <TextInput
    defaultValue="What did you inspect?"
    placeholder="Who inspected freshness?"
    testID="bananaChef"
    value="I inspected freshie"
  />
  <TextInput
    defaultValue="What banana?"
  />
  <TextInput
    defaultValue="hello"
    value=""
  />
  <MyButton
    onPress={[Function changeFresh]}
    type="primary"
  >
    Change freshness!
  </MyButton>
  <Text
    testID="duplicateText"
  >
    First Text
  </Text>
  <Text
    testID="duplicateText"
  >
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;

exports[`debug: with message 1`] = `
"my custom message

<View>
  <Text>
    Is the banana fresh?
  </Text>
  <Text
    testID="bananaFresh"
  >
    not fresh
  </Text>
  <TextInput
    placeholder="Add custom freshness"
    testID="bananaCustomFreshness"
    value="Custom Freshie"
  />
  <TextInput
    defaultValue="What did you inspect?"
    placeholder="Who inspected freshness?"
    testID="bananaChef"
    value="I inspected freshie"
  />
  <TextInput
    defaultValue="What banana?"
  />
  <TextInput
    defaultValue="hello"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function onBlur]}
    onClick={[Function onClick]}
    onFocus={[Function onFocus]}
    onResponderGrant={[Function onResponderGrant]}
    onResponderMove={[Function onResponderMove]}
    onResponderRelease={[Function onResponderRelease]}
    onResponderTerminate={[Function onResponderTerminate]}
    onResponderTerminationRequest={[Function onResponderTerminationRequest]}
    onStartShouldSetResponder={[Function onStartShouldSetResponder]}
  >
    <Text>
      Change freshness!
    </Text>
  </View>
  <Text
    testID="duplicateText"
  >
    First Text
  </Text>
  <Text
    testID="duplicateText"
  >
    Second Text
  </Text>
  <Text>
    0
  </Text>
</View>"
`;
